src/app/modules/form/admin-form/admin-form.controller.ts

```diff
@@ -2533,3 +2534,123 @@ export const handleGetFreeSmsCountForFormAdmin: ControllerHandler<
       })
   )
 }
+
+/**
+ * Handler for PUT /:formId/twilio.
+ * @security session
+ *
+ * @returns 200 with twilio credentials succesfully updated
+ * @returns 400 with twilio credentials are invalid
+ * @returns 401 when user is not logged in
+ * @returns 403 when user does not have permissions to update the form
+ * @returns 404 when form to update cannot be found
+ * @returns 422 when id of user who is updating the form cannot be found
+ * @returns 500 when database error occurs
+ */
+export const handleUpdateTwilio: ControllerHandler<
+  { formId: string },
+  unknown,
+  TwilioCredentials
+> = (req, res) => {
+  const { formId } = req.params
+  const twilioCredentials = req.body
+
+  const sessionUserId = (req.session as AuthedSessionData).user._id
+
+  return UserService.getPopulatedUserById(sessionUserId)
+    .andThen((user) =>
+      AuthService.getFormAfterPermissionChecks({
+        user,
+        formId,
+        level: PermissionLevel.Write,
+      }),
+    )
+    .andThen((retrievedForm) => {
+      const { msgSrvcName } = retrievedForm
+
+      return msgSrvcName
+        ? AdminFormService.updateTwilioCredentials(
+            msgSrvcName,
+            twilioCredentials,
+          )
+        : AdminFormService.createTwilioCredentials(
+            twilioCredentials,
+            retrievedForm,
+          )
+    })
+    .map(() =>
+      res
+        .status(StatusCodes.OK)
+        .json({ message: 'Successfully updated Twilio credentials' }),
+    )
+    .mapErr((error) => {
+      logger.error({
+        message: 'Error occurred when updating twilio credentials',
+        meta: {
+          action: 'handleUpdateTwilio',
+          ...createReqMeta(req),
+          userId: sessionUserId,
+          formId,
+          twilioCredentials,
+        },
+        error,
+      })
+      const { errorMessage, statusCode } = mapRouteError(error)
+      return res.status(statusCode).json({ message: errorMessage })
+    })
+}
+
+/**
+ * Handler for DELETE /:formId/twilio.
+ * @security session
+ *
+ * @returns 200 with twilio credentials succesfully updated
+ * @returns 401 when user is not logged in
+ * @returns 403 when user does not have permissions to update the form
+ * @returns 404 when form to delete credentials cannot be found
+ * @returns 422 when id of user who is updating the form cannot be found
+ * @returns 500 when database error occurs
+ */
+export const handleDeleteTwilio: ControllerHandler<{ formId: string }> = (
```

Comment:
should we add Joi validators in front of this controller handler fn? So the burden of twilio validity can be caught earlier before invocation of the service functions

Reply:
I don't think there's a need to Joi validator for this, since we only need the `formId` to carry out the functionalities :D

Reply:
we should still have Joi validator for validating the body, no? What happens if someone POSTs rubbish objects to the endpoint? We should return errors as soon as possible